### DISCUSSION TITLE
Composite + virtualization, how to sync them?

### DISCUSSION THREAD

**Discussion by cloud-walker** [ariakit/ariakit#3960] [Questions] [answered]
Hi! I'm trying to achieve a virtualized Select + Combobox and based on previous discussions (#1930, #3812), I understand that:

1. there is a SelectRenderer experiment inside ariakit, that is still experimental
2. one of the external virtualization solutions suggested seems to be [virtua](https://github.com/inokawa/virtua)

I've used virtua (here the example: https://stackblitz.com/edit/vitejs-vite-8vgxmj?file=src%2FAirplanesVSelect.tsx) and it works fine but the keyboard events are broken (pressing home or end keys not working as the items does not exists in the DOM because of the virtualization, I suppose).

Is there a way to sync the Composite behaviour of ariakit with the Virtualizer of virtua? (I guess if its there a way it can be used in other virtuatlization solutions too).

Searching in the docs I see that there is the [setActiveId](https://ariakit.org/reference/use-composite-store#setactiveid) function, that maybe I can use to link it to the scrollToIndex function of the virtua Virtualizer, but I suspect is not enough.

If there is not currently a way, what about exposing a primitive API to hook into Composite events?

**Comment by cloud-walker** [comment, marked as answer]
I've found a solution without using ariakit primitives, but leveraging the virtua's Virtualizer `keepMounted` prop

```tsx
function VItems() {
  return (
    <Virtualizer
      count={airplanes.length}
      overscan={10}
      keepMounted={[0, airplanes.length - 1]} // this keeps first and last items always mounted
    >
      {(index) => {
        const airplane = airplanes[index] ?? raise('Airplane not found');
        return (
          <A.SelectItem
            key={airplane.iataTypeCode}
            render={<A.ComboboxItem />}
            className="select-item truncate"
            value={airplane.name}
          />
        );
      }}
    </Virtualizer>
  );
}
```

**Reply by diegohaz** [reply]
That's the correct solution. In our `SelectRenderer`, we also keep the first and last enabled items mounted.
